scripts/e2e-source-install.py

```
@@ -0,0 +1,360 @@
+        _validate_owned(tombstone, slot_name=root.name)
+    if _entry_exists(root) and _entry_exists(tombstone):
+        _fail("canonical and retired install homes both exist")
+    if _entry_exists(tombstone):
+        # A killed recursive retirement can leave root-owned descendants.
+        # Restore the still-marked tree to the canonical slot so the workflow's
+        # next runner-cleanup invocation repairs it before prepare/cleanup.
+        os.rename(tombstone, root)
+        _validate_owned(root)
+
+
+def _preserve_persistent_build_tools(root: Path, persistent_bin: Path) -> Path | None:
```

Ruff (0.16.1)

[warning] Too many branches (15 > 12)

(PLR0912)